cpp/include/rapidsmpf/buffer/content_description.hpp

```diff
@@ -0,0 +1,135 @@
+     *     },
+     *     true // spillable
+     * };
+     * @endcode
+     */
+    template <std::ranges::input_range Range>
+        requires std::convertible_to<
+            std::ranges::range_value_t<Range>,
+            std::pair<MemoryType, std::size_t>>
+    explicit ContentDescription(Range&& sizes, bool spillable) : spillable(spillable) {
+        content_sizes.fill(0);
+        for (auto&& [mem_type, size] : sizes) {
+            auto idx = static_cast<std::size_t>(mem_type);
+            if (idx < content_sizes.size()) {
+                content_sizes[idx] = size;
+            }
+        }
+    }
+
+    /**
+     * @brief Construct from an initializer list of (MemoryType, size) pairs.
+     *
+     * Unspecified memory types becomes zero.
+     *
+     * @param pairs Initializer list of (MemoryType, size) pairs.
+     * @param spillable Whether the content are spillable.
+     */
+    ContentDescription(
+        std::initializer_list<std::pair<MemoryType, std::size_t>> pairs, bool spillable
+    )
+        : ContentDescription(
+              std::span<std::pair<MemoryType, std::size_t> const>(
+                  pairs.begin(), pairs.size()
+              ),
+              spillable
+          ) {}
+
+    /**
+     * @brief Construct a description with all sizes zero and a given spillability.
+     *
+     * Useful when you need to build a content description iteratively.
+     *
+     * @param spillable Whether the content are spillable.
+     */
+    ContentDescription(bool spillable) : ContentDescription({{}}, spillable) {}
+
+    /**
+     * @brief Access (read/write) the size for a specific memory type.
+     *
+     * @param mem_type The memory type entry to access.
+     * @return Reference to the size (in bytes) for the given memory type.
+     */
+    std::size_t& content_size(MemoryType mem_type) noexcept {
+        return content_sizes[static_cast<std::size_t>(mem_type)];
+    }
+
+    /// @brief Per memory-type content sizes, in bytes. Unspecified entries are zero.
+    std::array<std::size_t, MEMORY_TYPES.size()> content_sizes = {};
+
+    /// @brief Whether the content can be spilled.
+    bool spillable{false};
+};
+
+/**
+ * @brief Generate a host-only `ContentDescription` for numeric or string values.
+ *
+ * @tparam T Must be an integral type, a floating-point type, or `std::string`.
+ * @param obj The object's content to describe.
+ * @return A new content description.
+ */
+template <typename T>
+    requires std::is_integral_v<std::remove_cvref_t<T>>
+             || std::is_floating_point_v<std::remove_cvref_t<T>>
+             || std::same_as<std::remove_cvref_t<T>, std::string>
+ContentDescription get_content_description(T const& obj) {
+    using U = std::remove_cvref_t<T>;
+    std::size_t size;
+    if constexpr (std::same_as<U, std::string>) {
+        size = obj.size();
+    } else {
+        size = sizeof(U);
+    }
+    return ContentDescription{{{MemoryType::HOST, size}}, true};
+}
```

Comment:
Why do we need this? These things will never be on device, right? (certainly not the `std::string`).

Reply:
Everything that goes into a message must have a `ContentDescription`